jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java

```diff
@@ -20,131 +19,54 @@ public JdbcTemplate(final DataSource dataSource) {
         this.dataSource = dataSource;
     }
 
-    public void update(final String sql, final Object... args) {
-        Connection conn = null;
-        PreparedStatement pstmt = null;
-        try {
-            conn = dataSource.getConnection();
-            pstmt = conn.prepareStatement(sql);
-
-            log.debug("query : {}", sql);
-            setParameters(args, pstmt);
-
-            pstmt.executeUpdate();
-        } catch (SQLException e) {
-            log.error(e.getMessage(), e);
-            throw new RuntimeException(e);
-        } finally {
-            try {
-                if (pstmt != null) {
-                    pstmt.close();
-                }
-            } catch (SQLException ignored) {
+    public <T> T query(
+            final String sql,
+            final PreparedStatementSetter pss,
+            final ResultSetExtractor<T> rse
+    ) {
+        return execute(sql, (pstmt) -> {
+            pss.setValues(pstmt);
+            try (final ResultSet rs = pstmt.executeQuery()) {
+                return rse.extractData(rs);
             }
+        });
+    }
 
-            try {
-                if (conn != null) {
-                    conn.close();
-                }
-            } catch (SQLException ignored) {
-            }
-        }
+    public <T> List<T> query(final String sql, final RowMapper<T> rowMapper, final Object... args) {
+        return query(sql, new ArgumentPreparedStatementSetter(args), new RowMapperResultSetExtractor<>(rowMapper));
     }
 
     public <T> T queryForObject(final String sql, final RowMapper<T> rowMapper, final Object... args) {
-        Connection conn = null;
-        PreparedStatement pstmt = null;
-        ResultSet rs = null;
-        try {
-            conn = dataSource.getConnection();
-            pstmt = conn.prepareStatement(sql);
-            setParameters(args, pstmt);
-
-            rs = pstmt.executeQuery();
-
-            log.debug("query : {}", sql);
-
-            if (rs.next()) {
-                return rowMapper.mapRow(rs, 0);
-            }
+        final List<T> results = query(sql, new ArgumentPreparedStatementSetter(args),
+                new RowMapperResultSetExtractor<>(rowMapper));
+        if (results.isEmpty()) {
             return null;
```

Comment:
```suggestion
            throw new DataAccessException("Incorrect result size: expected 1, actual 0");
```
결과가 없는 경우 null을 반환해 NPE를 발생시키기보다 예외를 던지도록 하는 건 어떨까요?

Reply:
04a9ec8

실제 라이브러리를 참고해서 해당 부분 처리를 위한 CustomException을 추가했습니다 ⚡